File: src/main/java/bbarena/model/event/game/MatchStatusChangeEvent.java
package bbarena.model.event.game;

import bbarena.model.Arena;
import bbarena.model.Match;

import java.util.Date;

public class MatchStatusChangeEvent extends GameEvent {

    private Match.Status _prevMatchStatus;
    private Match.Status _nextMatchStatus;

    public MatchStatusChangeEvent(Match.Status newMatchStatus) {
        _nextMatchStatus = newMatchStatus;
    }

    @Override
    protected void doEvent(Arena arena) {
        _prevMatchStatus = arena.getMatch().getStatus();
        arena.getMatch().setStatus(_nextMatchStatus);
        if (Match.Status.STARTING == _nextMatchStatus) {
            arena.getMatch().setStart(new Date());
        } else if (Match.Status.FINISHED == _nextMatchStatus) {
            arena.getMatch().setEnd(new Date());
        }
    }

    @Override
    protected void undoEvent(Arena arena) {
        arena.getMatch().setStatus(_prevMatchStatus);
    }

    @Override
    public String getString() {
        StringBuilder res = new StringBuilder();
        res.append(getClass().getSimpleName())
                .append("[match:")
                .append(_nextMatchStatus)
                .append("]");
        return res.toString();
    }

    public Match.Status getMatchStatus() {
        return _nextMatchStatus;
    }
}

File: src/main/kotlin/bbarena/model/event/game/MatchStatusChangeEvent.kt
package bbarena.model.event.game

import bbarena.model.Arena
import bbarena.model.Match
import java.util.*

class MatchStatusChangeEvent(val matchStatus: Match.Status) : GameEvent() {

    private var prevMatchStatus: Match.Status? = null

    override fun doEvent(arena: Arena) {
        prevMatchStatus = arena.match.status
        arena.match.status = matchStatus

        // TODO: move in event listener
        if (Match.Status.STARTING == matchStatus) {
            arena.match.start = Date()
        } else if (Match.Status.FINISHED == matchStatus) {
            arena.match.end = Date()
        }
    }

    override fun undoEvent(arena: Arena) {
        arena.match.status = prevMatchStatus ?: Match.Status.STARTING
    }

    override fun getString(): String {
        val res = StringBuilder()
        res.append(javaClass.simpleName)
                .append("[match:")
                .append(matchStatus)
                .append("]")
        return res.toString()
    }
}
